### DISCUSSION TITLE
[QUESTION] - slick 2.1.0 working differently in scala 2.11.7 and scala 2.12.17

### DISCUSSION THREAD

**Discussion by sujit-baniya** [slick/slick#2654] [Questions]
I've following code that was working on scala 2.11 and slick 2.1.0.

```scala
case class User(
                   id: Int,
                   email: String,
                   hash: String,
                   firstName: String,
                   lastName: String,
                   user_active: Boolean,
                   added_by: Option[Int],
                   added_utc: Option[Timestamp],
                   gamma_tester: Boolean
               )

object User {
  implicit val codec = CodecJson.casecodec9(User.apply, User.unapply)(
    "id", "email", "hash", "firstName", "lastName", "user_active",
    "added_by", "added_utc", "gamma_tester")
  implicit def encode = codec.Encoder
  implicit def decode = codec.Decoder
  implicit val entityEncoder = jsonEncoderOf[User]
  implicit val optionEntityEncoder = jsonEncoderOf[Option[User]]
  implicit val listEntityEncoder = jsonEncoderOf[List[User]]
  
  implicit val entityDecoder = jsonOf[User]
  implicit val optionEntityDecoder = jsonOf[Option[User]]
  implicit val listEntityDecoder = jsonOf[List[User]]
}
 // OTHER file to get user
	def findUser(email: String) = {
		db.connect.withSession { implicit session =>
			(sql"SELECT user_uid, user_email_address, user_password, user_first_name, user_last_name, user_active,added_by,added_utc,gamma_tester FROM tbl_user WHERE user_email_address = $email".as[User]).first
		}
	}
```

Here I didn't require to use implicit GetResult.

```scala
case class User(
                   id: Int,
                   email: String,
                   hash: String,
                   firstName: String,
                   lastName: String,
                   user_active: Boolean,
                   added_by: Option[Int],
                   added_utc: Option[Timestamp],
                   gamma_tester: Boolean
               )

object User {
  implicit val codec = CodecJson.casecodec9(User.apply, User.unapply)(
    "id", "email", "hash", "firstName", "lastName", "user_active",
    "added_by", "added_utc", "gamma_tester")
  implicit def encode = codec.Encoder
  implicit def decode = codec.Decoder
  implicit val entityEncoder = jsonEncoderOf[User]
  implicit val optionEntityEncoder = jsonEncoderOf[Option[User]]
  implicit val listEntityEncoder = jsonEncoderOf[List[User]]
  
  implicit val entityDecoder = jsonOf[User]
  implicit val optionEntityDecoder = jsonOf[Option[User]]
  implicit val listEntityDecoder = jsonOf[List[User]]
  implicit val getUserResult = GetResult(r => User(r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<))
  implicit val getOptionUserResult = GetResult(r => Some(User(r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<)))
}
 // OTHER file to get user
	def findUser(email: String) = {
		db.connect.withSession { implicit session =>
			(sql"SELECT user_uid, user_email_address, user_password, user_first_name, user_last_name, user_active,added_by,added_utc,gamma_tester FROM tbl_user WHERE user_email_address = $email".as[User]).first
		}
	}
```

But here, I can't call the user and have to use `GetResult` in case class.

If I don't use `GetResult`, I get an error

```scala
No implicits found for parameter rconv: GetResult[User]
```

Please suggest why such behavior or if am I missing something

**Comment by nafg** [comment]
Slick 3 made big changes. It was release seven years ago. The maintainers from back then are long gone.

I suggest you read https://github.com/slick/slick/blob/c22cfdf5e14c73a81ed1e7d2a3d704ce14e5ac05/slick/src/sphinx/upgrade.rst#upgrade-from-21-to-30

**Reply by sujit-baniya** [reply]
@nafg I'm doing the incremental upgrade and started with Scala. Upgrading slick from 2.1.0 to 3, at the moment, is not possible as it has drastic changes.

**Reply by nafg** [reply]
I sympathize with your predicament. But working with a project that has been unmaintained for the better part of a decade is never easy.

Unfortunately I don't think you can get both versions on your classpath simultaneously, so if you have to do an incremental migration you might be better off porting to some other library that can be on the classpath together. Perhaps Quill would be an option. Otherwise you'd have to get the whole thing ported to modern Slick before it will compile. How big is this project?

**Reply by sujit-baniya** [reply]
By big, the project has been in production for the last 8 years with around 123 tables followed by raw queries, and the system was never upgraded.

**Reply by sujit-baniya** [reply]
I was just wondering why the same slick version has different results on different scala versions: 2.11.7 and 2.12.17

**Reply by nafg** [reply]
I don't know the answer.

But how many lines of Slick code are there?

**Reply by jkugiya** [reply]
>  No implicits found for parameter rconv: GetResult[User]

The problem appears to be that the compilation no longer passes, rather than that the runtime behavior has changed. I think the behavior of type inference may have changed.

```scala
 def findUser(email: String) = {
```

For public functions, it is recommended that explicit types be specified rather than leaving it to the compiler's type inference.
Does the same problem occur if you make the return value of `findUser` explicit?

**Reply by sujit-baniya** [reply]
I've one example of string interpolation in slick 2.1.0 which is giving error

```scala
  def chargeQuery(workItemId: Int,
                  date: YmdDate,
                  proc: String,
                  mod1: Option[String],
                  mod2: Option[String],
                  chargeCodeSelect: Option[String],
                  chargeType: ChargeType,
                  patient_status_uid: Option[Int]): Q[Unit,ChargeCode] = {
    val m1 = mod1.orZero
    val m2 = mod2.orZero
    sql"""
      SELECT charge_master_uid, client_internal_code, client_proc_desc, charge_amt, client_mnemonic, code
      FROM (
          SELECT work_item_uid, charge_type, MAX(effective_date) AS effective_date, patient_status_uid
          FROM tbl_charge_master
          WHERE work_item_uid = $workItemId
                AND charge_type = ${chargeType.chargeKey}
                AND effective_date <= ${date.toString}
        ) AS EffectiveDate
        JOIN tbl_charge_master USING (work_item_uid,charge_type,effective_date)
        LEFT JOIN tbl_work_item_patient_status ON tbl_charge_master.patient_status_uid = tbl_work_item_patient_status.patient_status_uid
          AND EffectiveDate.work_item_uid = tbl_work_item_patient_status.work_item_uid
        WHERE cpt_hcpcs_code IN (${proc+m1+m2}, ${proc+m2+m1}, ${proc+m1}, ${proc+m2}, $proc)
          AND client_internal_code <> ''
        ORDER BY LENGTH(cpt_hcpcs_code) DESC,
                 client_internal_code = ${chargeCodeSelect.orZero} DESC
        LIMIT 1
       """
      .as[ChargeCode]
  }
```

Error:
```scala
ChargeHelper.scala:66:48: 8 more arguments than can be applied to method sql: (param: (Int, String, String, String, String, String, String, String, String))(implicit pconv: scala.slick.jdbc.SetParameter[(Int, String, String, String, String, String, String, String, String)])scala.slick.jdbc.SQLInterpolationResult[(Int, String, String, String, String, String, String, String, String)]
[error]                 AND charge_type = ${chargeType.chargeKey}

```

**Reply by jkugiya** [reply]
There have been similar issues with this problem in the past, but it appears to have been resolved since 3.0.0.
https://github.com/slick/slick/issues/155

It seems that the Scala compiler treats tuples as normal arguments, so you get an error that there are extra arguments.
It was easy to reproduce with the following code.

```scala
import scala.slick.jdbc.StaticQuery.interpolation
import scala.slick.jdbc._

val i: Int = 1

val query1 = sql""" ${i} """ // ok
val query2 = sql""" ${i} , ${i}""" // ng
```

This problem seems difficult to solve by devising a way to use the library.
You may need to solve this problem yourself by referring to the macro-based implementation of Slick3 or later.
If that seems too difficult, you may have to use `? ` to write verbose code.

**Reply by nafg** [reply]
@jkugiya can you convert to an answer so we can mark it answered

**Reply by jkugiya** [reply]
https://github.com/slick/slick/discussions/2654#discussioncomment-5210403

**Comment by jkugiya** [comment]
There have been similar issues with this problem in the past, but it appears to have been resolved since 3.0.0.
https://github.com/slick/slick/issues/155

It seems that the Scala compiler treats tuples as normal arguments, so you get an error that there are extra arguments.
It was easy to reproduce with the following code.

```scala
import scala.slick.jdbc.StaticQuery.interpolation
import scala.slick.jdbc._

val i: Int = 1

val query1 = sql""" ${i} """ // ok
val query2 = sql""" ${i} , ${i}""" // ng
```

This problem seems difficult to solve by devising a way to use the library.
You may need to solve this problem yourself by referring to the macro-based implementation of Slick3 or later.
If that seems too difficult, you may have to use `? ` to write verbose code.

**Reply by sujit-baniya** [reply]
Maybe I could use blocking-slick https://github.com/gitbucket/blocking-slick change by changing some implementation
